File: solutions/lua/queen-attack/queen-attack.lua
function valid_range(n)
    return n >= 0 and n <= 7
end

function valid_position(pos)
    return valid_range(pos.row) and valid_range(pos.column)
end

function diagonal(pos, oth)
    return math.abs(pos.row - oth.row) == math.abs(pos.column - oth.column)
end

return function(pos)
    if not valid_position(pos) then error("invalid position") end

    pos.can_attack = function(oth)
        return pos.row == oth.row 
            or pos.column == oth.column
            or diagonal(pos, oth)
    end

    return pos
end

File: solutions/r/queen-attack/queen-attack.R
create <- function(row, col) {
    stopifnot(row >= 0 && row <= 7)
    stopifnot(col >= 0 && col <= 7)
    list(row = row, col = col)
}

can_attack <- function(queen1, queen2) {
    (queen1$row == queen2$row ||
        queen1$col == queen2$col ||
        abs(queen1$row - queen2$row) == abs(queen1$col - queen2$col))
}
